SPOC Recorder Improvements

mhils opened the pull request:
#### What type of PR is this?

/kind bug
/kind feature

#### What this PR does / why we need it:

This PR has two main parts:

1. Add a new `--no-start` flag that allows spoc to record profiles without driving the process execution. Processes are matched by executable name (spanning all mount namespaces), so this for example allows recording workloads in Docker containers.
2. When the AppArmor recorder terminates, wait until we have processed the process exit event for the right PID in Go. This ensures that no events are left unprocessed in the ebPF<->Go ring buffer.


#### Which issue(s) this PR fixes:

refs (but not fixes) https://github.com/kubernetes-sigs/security-profiles-operator/issues/718

#### Does this PR have test?

yes

#### Does this PR introduce a user-facing change?


```release-note
Add a new `--no-start` flag that allows spoc to record profiles without driving the process execution.
```

pjbgf commented:
/ok-to-test

ccojocar commented on internal/pkg/cli/recorder/options.go:
```diff
@@ -31,6 +31,7 @@ type Options struct {
 	typ            Type
 	outputFile     string
 	baseSyscalls   []string
+	noStart        bool
```
I would name it `noProcStart` to be more clear since these are the options for recorder itself. `noStart` sounds more like not starting the recorder.

ccojocar commented on internal/pkg/cli/recorder/recorder.go:
```diff
@@ -72,29 +78,46 @@ func (r *Recorder) Run() error {
+		ch := make(chan os.Signal, 1)
+		r.Notify(ch, os.Interrupt)
+		log.Print(WaitForSigIntMessage)
+		<-ch
```
What if the process has already exited before setting up the notification, is this going to be stuck here forever?

ccojocar replied:
Thanks for clarification. If returns immediately when the process is existed, there isn't a problem.

ccojocar commented on internal/pkg/daemon/bpfrecorder/bpfrecorder.go:
```diff
@@ -917,7 +913,13 @@ func (b *BpfRecorder) handleAppArmorEvents(apparmorEvents chan []byte) {
 			b.handleAppArmorCapabilityEvents(&apparmorEvent)
 		case uint8(probeTypeExit):
-			b.lockAppArmorRecording.Unlock()
+			d, _ := b.recordedExits.LoadOrStore(apparmorEvent.Pid, make(chan bool))
```
Is it possible that this might lead to double closing the channel, for instance if the channel already exist in the map? 

I believe is just one exit event coming, but maybe make senses to check this just to be sure.

ccojocar reviewed (commented):
It looks good, just left a few minor comments.

mhils commented on internal/pkg/cli/recorder/recorder.go:
```diff
@@ -103,40 +126,51 @@ func (r *Recorder) Run() error {
 				}
 
-				syscalls = append(syscalls, name)
+				syscallsMap[name] = true
```
With `--no-start`, we are taking all mount namespaces into account (we match on the process name only). So this is using a map as a set to avoid duplicates.

ccojocar replied:
Thanks for clarifications, it makes sense then.

mhils commented on internal/pkg/cli/recorder/recorder.go:
```diff
@@ -103,40 +126,51 @@ func (r *Recorder) Run() error {
-			return fmt.Errorf("build profile: %w", err)
-		}
+	// map -> slice
+	syscalls := make([]string, len(syscallsMap))
```
I'll add a clarifying comment. 👍

ccojocar replied:
This sounds good.

mhils commented:
Thanks for the quick feedback @ccojocar! I've addressed all feedback, but we're still hitting test failures in CI that I can neither explain yet nor reproduce locally 😛. Setting this to draft for now until I have figured things out. :)

ccojocar commented:
@mhils There are also some lint warnings, and I think there is a syscall which is removed from the profile.

mhils commented:
Fixed! Turns out sudo does not like forwarding SIGINT in some cases, found a workaround for the test. :)

mhils commented:
/retest

mhils commented:
/retest

mhils commented:
There are still two CI failures, but I believe those are flaky CI and not related to this PR. :)
